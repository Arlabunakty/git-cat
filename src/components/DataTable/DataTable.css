.table-header {
  display: flex;
  vertical-align: top;
  text-align: left;
  margin: 0;
  padding: 1em 1em;
}

.table-name {
  font-weight: bold;
}

.icon-rtl {
  padding-right: 25px;
  background: url("./../../assets/101791-200.png") no-repeat right;
  background-size: 20px;
  width: 8em;
  border-color: #8b8b8b;
  border-radius: 0.25em;
}
.left-header {
  display: flex;
  flex-direction: column;
}
.right-header {
  margin-left: auto;
  order: 2;
}
.table {
  width: 90%;
  margin: 0 1.8em;
  border: 0px;
  clear: both;
  border-spacing: 0;
  border-collapse: separate;
}

.th,
.td {
  font-weight: normal;
  background: #ffffff;
  color: black;
  height: 3em;
}

.td {
  text-align: center;
}

table .th:nth-child(1) {
  border-radius: 0 0 0 1em;
}
table .th:nth-last-child(1) {
  border-radius: 0 0 1em 0;
}

table .th.first-row {
  border-radius: 1em 1em 0 0;
}

table .td:nth-child(1) {
  border-radius: 1em 0 0 1em;
}
table .td:nth-last-child(1) {
  border-radius: 0 1em 1em 0;
}
.tbody:before {
  line-height: 0.25em;
  content: "\200C";
  display: block;
}
.td {
  border-bottom-style: dashed;
  border-bottom-color: #f3f7fa;
  max-width: 15em;
  height: 5em;
}
