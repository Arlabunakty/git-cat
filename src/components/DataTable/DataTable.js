import React, { useState } from "react";
import "./DataTable.css";

const DataTable = ({
  title,
  description,
  headers,
  data,
  testid,
  isLoading,
  errorMessage,
}) => {
  if (isLoading) {
    return <p>Fetching...</p>;
  }
  if (errorMessage != null) {
    return <p>{errorMessage}</p>;
  }
  const [search, setSearch] = useState("");
  const originalData = data;
  const searchableHeaders = headers.filter((header) => header.searchable);

  let resultData = originalData;
  if (search !== "") {
    resultData = originalData.filter((row) =>
      searchableHeaders.some((header) =>
        (row[header.propertyName] || "").toLowerCase().includes(search)
      )
    );
  }
  return (
    <div className="table-container" data-testid={testid}>
      <table className="table">
        <thead className="thead">
          <tr className="tr">
            <th scope="col" colSpan={headers.length} className="th first-row">
              <div className="table-header">
                <div className="left-header">
                  <span data-testid="table-name" className="table-name">
                    {title}
                  </span>
                  <span data-testid="table-description">{description}</span>
                </div>
                <div className="right-header">
                  <input
                    data-testid="search-input"
                    type="text"
                    className="icon-rtl"
                    placeholder="Search"
                    onChange={(e) => setSearch(e.target.value.toLowerCase())}
                  />
                </div>
              </div>
            </th>
          </tr>
          <tr>
            {headers.map((h) => (
              <th key={h.name} scope="col" className="th column-name">
                {h.name}
              </th>
            ))}
          </tr>
        </thead>
        <tbody className="tbody">
          {resultData.map((row, i) => renderRow(row, i, headers))}
        </tbody>
      </table>
    </div>
  );
};

function renderCell(header, j, row) {
  const value = row[header.propertyName];
  return (
    <td className="td" key={j}>
      {typeof value === "boolean" ? "" + value : value}
    </td>
  );
}
function renderRow(row, i, headers) {
  return (
    <tr className="tr table-row" key={i}>
      {headers.map((header, j) => renderCell(header, j, row))}
    </tr>
  );
}

export default DataTable;
