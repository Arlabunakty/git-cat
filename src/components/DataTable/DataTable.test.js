import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import DataTable from "./DataTable";

const headers = [
  { name: "Col1", propertyName: "col1", searchable: true },
  { name: "Col2", propertyName: "col2" },
  { name: "BOOLEAN", propertyName: "col3" },
  { name: "NULL", propertyName: "col4" },
];
const data = [
  { col1: "TextCol1", col2: "textcol2", col3: true, col4: null },
  { col1: null, col2: "another", col3: false, col4: undefined },
];

it("renders empty table", async () => {
  const { container } = render(
    <DataTable
      title="TableName"
      description="TableDescription"
      headers={[]}
      data={[]}
    />
  );

  expect((await screen.findByTestId("table-name")).textContent).toEqual(
    "TableName"
  );
  expect((await screen.findByTestId("table-description")).textContent).toEqual(
    "TableDescription"
  );
  expect(await screen.findByTestId("search-input")).toBeInTheDocument();
  expect(container.querySelectorAll(".table-row").length).toEqual(0);
});

it("renders table with data", async () => {
  const { container } = render(
    <DataTable
      title="TableNameWith2Columns"
      description="TableDescriptionFOrTableWith2Columns"
      headers={headers}
      data={data}
    />
  );

  expect((await screen.findByTestId("table-name")).textContent).toEqual(
    "TableNameWith2Columns"
  );
  expect((await screen.findByTestId("table-description")).textContent).toEqual(
    "TableDescriptionFOrTableWith2Columns"
  );
  expect(await screen.findByTestId("search-input")).toBeInTheDocument();
  const rows = container.querySelectorAll(".table-row");
  expect(rows.length).toEqual(2);
  const cells = rows[0].querySelectorAll(".td");
  expect(cells.length).toEqual(4);
  expect(cells[0].textContent).toEqual("TextCol1");
  expect(cells[1].textContent).toEqual("textcol2");
  expect(cells[2].textContent).toEqual("true");
  expect(cells[3].textContent).toEqual("");

  const headerCols = container.querySelectorAll(".th.column-name");
  expect(headerCols.length).toEqual(4);
  expect(headerCols[0].textContent).toEqual("Col1");
  expect(headerCols[1].textContent).toEqual("Col2");
  expect(headerCols[2].textContent).toEqual("BOOLEAN");
  expect(headerCols[3].textContent).toEqual("NULL");
});

it.each([
  {
    search: "TextCol3",
    expectedNumOfRows: 0,
  },
  {
    search: "true",
    expectedNumOfRows: 0,
  },
  {
    search: "a",
    expectedNumOfRows: 0,
  },
  {
    search: "TextCol1",
    expectedNumOfRows: 1,
  },
  {
    search: "TextCol",
    expectedNumOfRows: 1,
  },
  {
    search: "TextCo",
    expectedNumOfRows: 1,
  },
  {
    search: "col",
    expectedNumOfRows: 1,
  },
  {
    search: "1",
    expectedNumOfRows: 1,
  },
  {
    search: "text",
    expectedNumOfRows: 1,
  },
])("filter table by text input", async (test) => {
  const { container } = render(
    <DataTable
      title="TableNameWith2Columns"
      description="TableDescriptionFOrTableWith2Columns"
      headers={headers}
      data={data}
    />
  );
  const search = await screen.findByTestId("search-input");

  fireEvent.change(search, { target: { value: test.search } });

  expect(container.querySelectorAll(".table-row").length).toEqual(
    test.expectedNumOfRows
  );
});
