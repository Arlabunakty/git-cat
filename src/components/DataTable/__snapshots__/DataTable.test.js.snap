// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders avatar column 1`] = `
<div
  className="table-container"
>
  <table
    className="table"
  >
    <thead
      className="thead"
    >
      <tr
        className="tr"
      >
        <th
          className="th first-row"
          colSpan={2}
          scope="col"
        >
          <div
            className="table-header"
          >
            <div
              className="left-header"
            >
              <span
                className="table-name"
                data-testid="table-name"
              />
              <span
                data-testid="table-description"
              />
            </div>
            <div
              className="right-header"
            >
              <input
                className="icon-rtl"
                data-testid="search-input"
                onChange={[Function]}
                placeholder="Search"
                type="text"
              />
            </div>
          </div>
        </th>
      </tr>
      <tr>
        <th
          className="th column-name"
          scope="col"
        >
          Owner
        </th>
        <th
          className="th column-name"
          scope="col"
        >
          Owner column
        </th>
      </tr>
    </thead>
    <tbody
      className="tbody"
    >
      <tr
        className="tr table-row"
      >
        <td
          className="td"
        >
          <div
            className="avatar-cell avatar-cell-row"
          >
            <img
              alt="avatar"
              className="avatar-img"
              src="avatar_url"
            />
            <div
              className="avatar-text"
            >
              <span>
                Hard coded name
              </span>
              <span />
            </div>
          </div>
        </td>
        <td
          className="td"
        >
          <div
            className="avatar-cell avatar-cell-column"
          >
            <img
              alt="avatar"
              className="avatar-img"
              src="avatar_url"
            />
            <div
              className="avatar-text"
            >
              <span>
                Hard coded name
              </span>
              <span />
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
